### PR TITLE
feat: Enable native Parquet writer in Spark SQL tests

### LINKED ISSUE #3209: Run Spark tests with Comet native writer
### What is the problem the feature request solves?

Comet native writer works for basic ops, however to move forward and keep compatibility with Spark we need to run Spark writer and partition discovery tests with Comet writer enabled first to make sure we have a solid reliable base for Comet native writer changes

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by coderfender**
## Which issue does this PR close?
 
Closes: https://github.com/apache/datafusion-comet/issues/3209 

3 kinds of tests have been identified as incompatible for native writer : 

1. We have 12 tests in SQL suite which failed / exposed corner cases with Comet writers ( Insert Overwrite / CTAS statements)
2. SQL tests for the `charvarcha.sql` files
3. Parquet metadata encoding tests (Spark writes metadata such as spark version while comet doesnt)

## Rationale for this change

Now that we have a native comet writer, this PR enables spark tests while ignoring the ones which fail 

## What changes are included in this PR?

Diff file changes to enable comet native writer on spark tests

### CODE REVIEW

File: .github/workflows/spark_sql_test.yml [hunk 1]
```diff
@@ -156,7 +156,7 @@ jobs:
         run: |
           cd apache-spark
           rm -rf /root/.m2/repository/org/apache/parquet # somehow parquet cache requires cleanups
-          ENABLE_COMET=true ENABLE_COMET_ONHEAP=true ${{ matrix.config.scan-env }} ENABLE_COMET_LOG_FALLBACK_REASONS=${{ github.event.inputs.collect-fallback-logs || 'false' }} \
+          ENABLE_COMET=true ENABLE_COMET_ONHEAP=true ENABLE_COMET_WRITER=true ${{ matrix.config.scan-env }} ENABLE_COMET_LOG_FALLBACK_REASONS=${{ github.event.inputs.collect-fallback-logs || 'false' }} \
```
Comment (comphead) [inline .github/workflows/spark_sql_test.yml]: we also need to enable 
```
--conf spark.comet.operator.DataWritingCommandExec.allowIncompatible=true
```

Comment (coderfender) [reply .github/workflows/spark_sql_test.yml]: Sure

**Comment by coderfender** [comment]
@comphead  , We have bunch of failures largely due to hitting edge cases with comet writer (CTAS , Insert Overwrite which is likely a bug) . I am thinking of two possible options here
1. Keep the changes minimal and ignore the failing I/O tests for now 
2. Create a separate comet writer suite which ignores all the tests relevant to Comet native writer

File: .github/workflows/spark_sql_test.yml (hunk 1 above)
Comment (coderfender) [reply .github/workflows/spark_sql_test.yml]: We currently dont have the config available in spark tests through `ENABLE_COMET_WRITER=true` . Perhaps you meant enabling them in the benchmark suite  TPCH / TPCDS  as well ?

**Comment by coderfender** [comment]
@comphead  please kickoff CI whenever you get a chance

**Comment by coderfender** [comment]
My diff files for Spark versions 3.4 and 3.5 missed disabling comet writer for CTAS statements (while it was set for Spark 4) causing above test failures . However, there are only 2 failed tests down from 13

**Comment by coderfender** [comment]
Seems like there were issues applying diff file for Spark 4

**Comment by coderfender** [comment]
@comphead all the tests passed .Please take a look whenever you get a chance

**Comment by coderfender** [comment]
All tests passed and the following tests are ignored : 

```
  Common across all Spark versions (3.4, 3.5, 4.0) :

  1. empty file should be skipped while write to file
  2. INSERT INTO TABLE - complex type but different names
  3. Insert overwrite table command should output correct schema: basic
  4. parquet timestamp conversion
  5. SPARK-29174 Support LOCAL in INSERT OVERWRITE DIRECTORY to data source
  6. SPARK-33901: ctas should should not change table's schema
  7. SPARK-37160: CREATE TABLE AS SELECT with CHAR_AS_VARCHAR
  8. SPARK-38336 INSERT INTO statements with tables with default columns: positive tests
  9. SPARK-38811 INSERT INTO on columns added with ALTER TABLE ADD COLUMNS: Positive tests
  10. SPARK-43071: INSERT INTO from queries whose final operators are not projections
  11. write path implements onTaskCommit API correctly
  12. Write Spark version into Parquet metadata

Additional tests which had to be ignored for spark 4 : 

  1. ctas with union
  2. SPARK-48817: test multi inserts

  Total: 12 tests (Spark 3.4/3.5) | 14 tests (Spark 4.0)
  ```

File: dev/diffs/3.4.3.diff [hunk 2]
```diff
@@ -1872,6 +1883,62 @@ index 593bd7bb4ba..32af28b0238 100644
...
+ import org.apache.spark.sql.catalyst.util.CharVarcharUtils
+ import org.apache.spark.sql.connector.catalog.InMemoryPartitionTableCatalog
+ import org.apache.spark.sql.internal.SQLConf
+@@ -112,7 +112,8 @@ trait CharVarcharDDLTestBase extends QueryTest with SQLTestUtils {
+     }
+   }
+ 
+-  test("SPARK-33901: ctas should should not change table's schema") {
++  test("SPARK-33901: ctas should should not change table's schema",
++    IgnoreComet("comet native writer does not support empty dir / table creation yet")) {
```
Comment (andygrove) [inline dev/diffs/3.4.3.diff]: Could you link to the specific GitHub issues in the `IgnoreComet` annotations?

**Comment by coderfender** [comment]
Fixing merge conflicts for the diff files :)

**Comment by coderfender** [comment]
Created github issues for all the failing tests and added links in IgnoreComet annotations

File: .github/workflows/spark_sql_test.yml [hunk 3]
```diff
@@ -155,7 +155,7 @@ jobs:
         run: |
           cd apache-spark
           rm -rf /root/.m2/repository/org/apache/parquet # somehow parquet cache requires cleanups
-          NOLINT_ON_COMPILE=true ENABLE_COMET=true ENABLE_COMET_ONHEAP=true COMET_PARQUET_SCAN_IMPL=${{ matrix.config.scan-impl }} ENABLE_COMET_LOG_FALLBACK_REASONS=${{ github.event.inputs.collect-fallback-logs || 'false' }} \
+          NOLINT_ON_COMPILE=true ENABLE_COMET=true ENABLE_COMET_ONHEAP=true ENABLE_COMET_WRITER=true COMET_PARQUET_SCAN_IMPL=${{ matrix.config.scan-impl }} ENABLE_COMET_LOG_FALLBACK_REASONS=${{ github.event.inputs.collect-fallback-logs || 'false' }} \
```
Comment (comphead) [inline .github/workflows/spark_sql_test.yml]: I would suggest to have a separate CI pipeline for writer for now which can be run on demand

File: dev/diffs/3.4.3.diff (hunk 2 above)
Comment (coderfender) [reply dev/diffs/3.4.3.diff]: Thank you . I did annotations and pushed a commit . However, per @comphead  's review, I made this PR just to copy the failing tests into the `CometParquetWriterSuite` and created github issues for each of the failing tests in spark 3x , spark 4x. 

The plan is to create a CI pipeline (which doesn't interfere with main comet workflows but runs a separate test suite with comet writer enabled and a listener attached to make sure we are running the tests through native writer

**Comment by mbutrovich** [comment]
Rather than ignoring tests in diffs, are we able to identify these scenarios at planning time and write fallbacks?

**Comment by coderfender** [comment]
@mbutrovich  , thank you for the comment. These tests were essentially exposing bugs with our writer which are not caught by our current fallbacks

  1. Empty file/dir creation
  empty file should be skipped while write to file
  SPARK-33901: ctas should not change table's schema
  SPARK-37160: CREATE TABLE AS SELECT with CHAR_AS_VARCHAR

  2. Insert overwrite bugs (5 tests)
  Insert overwrite table command should output correct schema: basic
  SPARK-29174 Support LOCAL in INSERT OVERWRITE DIRECTORY
  SPARK-38336 INSERT INTO with default columns
  SPARK-38811 INSERT INTO on ALTER TABLE ADD COLUMNS
  SPARK-43071: INSERT INTO from non-projection queries

  3. Other
  parquet timestamp conversion (timestamp96)
  INSERT INTO TABLE - complex type but different names
  Write Spark version into Parquet metadata
  write path implements onTaskCommit API correctly

I created separate issues for each failing tests, annotated them in the diffs . However, per my discussion with @comphead  , this PR is just the first step to copy those tests into comet test suite to lay foundation for a future comet writer only test suite in CI .

**Review by comphead — COMMENTED** [review]
Thanks @coderfender great to identify issues for the writer.
One question how is it planned to check native writer has been called? 
You might consider the option to call `writeWithCometNativeWriteExec` or look into how SparkListener used inside

**Comment by coderfender** [comment]
Great suggestion @comphead . I initially added a custom listerner to check that the native writer is called. However, I figured it might be cleaner to limit the scope of this PR to just tag the failing tests and follow up with more work to build on this. Do you think it is a good idea to add that verification in this PR or would you prefer the listener changes in a different one ?

**Comment by coderfender** [comment]
nevermind I think I got a better idea after this PR : https://github.com/apache/datafusion-comet/pull/3480

File: spark/src/test/scala/org/apache/comet/parquet/CometParquetWriterSuite.scala [hunk 4]
```diff
@@ -466,6 +737,11 @@ class CometParquetWriterSuite extends CometTestBase {
       s"Expected exactly one CometNativeWriteExec in the plan, but found $nativeWriteCount:\n${plan.treeString}")
   }
 
+  private def checkCometNativeWriter(op: => Unit): Unit = {
+    val plan = captureWritePlan(_ => op, "")
+    assertHasCometNativeWriteExec(plan)
+  }
+
```
Comment (coderfender) [inline spark/src/test/scala/org/apache/comet/parquet/CometParquetWriterSuite.scala:744]: @comphead , I wrote this simple util to match our other test patterns to verify that comet native writer is being used.

File: spark/src/test/scala/org/apache/comet/parquet/CometParquetWriterSuite.scala [hunk 5]
```diff
@@ -377,6 +378,276 @@ class CometParquetWriterSuite extends CometTestBase {
...
+    assume(isSpark40Plus)
+    withSQLConf(
+      CometConf.COMET_NATIVE_PARQUET_WRITE_ENABLED.key -> "true",
+      CometConf.COMET_EXEC_ENABLED.key -> "true",
+      CometConf.getOperatorAllowIncompatConfigKey(classOf[DataWritingCommandExec]) -> "true") {
+      withTable("t1", "t2") {
+        sql("CREATE TABLE t1(a INT) USING parquet")
+        sql("CREATE TABLE t2(a INT) USING parquet")
+        checkCometNativeWriter {
+          sql("FROM (SELECT 1 AS a) src INSERT INTO t1 SELECT a INSERT INTO t2 SELECT a")
```
Comment (comphead) [inline spark/src/test/scala/org/apache/comet/parquet/CometParquetWriterSuite.scala:643]: is it valid sql?

File: spark/src/test/scala/org/apache/comet/parquet/CometParquetWriterSuite.scala [hunk 6]
```diff
@@ -466,6 +737,11 @@ class CometParquetWriterSuite extends CometTestBase {
       s"Expected exactly one CometNativeWriteExec in the plan, but found $nativeWriteCount:\n${plan.treeString}")
   }
 
+  private def checkCometNativeWriter(op: => Unit): Unit = {
```
Comment (comphead) [inline spark/src/test/scala/org/apache/comet/parquet/CometParquetWriterSuite.scala]:
```suggestion
  private def assertCometNativeWriterCaptured(op: => Unit): Unit = {
```

File: spark/src/test/scala/org/apache/comet/parquet/CometParquetWriterSuite.scala (hunk 5 above)
Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/parquet/CometParquetWriterSuite.scala:643]: Thank you for the comment @comphead. I had the same surprise as well when I followed the same approach as the test we are porting to comet test suites . Multi table inserts are supported in spark (carried over approach from hive syntax). Additional details here : https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DML#LanguageManualDML-Syntax.1

Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/parquet/CometParquetWriterSuite.scala:643]: Corresponding spark test :
```
  test("SPARK-48817: test multi inserts") {
    withTable("t1", "t2", "t3") {
      createTable("t1", Seq("i"), Seq("int"))
      createTable("t2", Seq("i"), Seq("int"))
      createTable("t3", Seq("i"), Seq("int"))
      sql(s"INSERT INTO t1 VALUES (1), (2), (3)")
      val df = sql(
        """
          |FROM (select /*+ REPARTITION(3) */ i from t1)
          |INSERT INTO t2 SELECT i
          |INSERT INTO t3 SELECT i
          |""".stripMargin
      )
     
```
Not sure if I am missing anything here

Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/parquet/CometParquetWriterSuite.scala:643]: Related links : https://github.com/apache/spark/blob/v3.5.1/sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4#L444 

https://github.com/reata/sqllineage/issues/596
